packages/tailwindcss/src/plugin-api.ts

```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
         })
       }
     },
+    theme(path) {
+      let cssValue = readFromCss(designSystem.theme, path)
+
+      if (typeof cssValue !== 'object') {
+        return cssValue
+      }
+
+      let result = resolvedConfig.theme ?? {}
+
+      // 1. Convert `path` to an array of parts (segment)
+      // 2. Find the "obeject" that corresponds to the path
+      // 3. Deep merge the object with the resolved theme
+      let parts = segment(path, '.')
+
+      let obj = get(result, parts) ?? {}
+
+      if (typeof obj === 'object') {
+        deepMerge(obj, [cssValue], (a, b) => b)
+        set(result, parts, obj)
+      } else {
+        // This case happens when the CSS theme produces an object for a given
+        // keypath but the v3-style config does not. However we still want the
+        // CSS to "win" in this case so we overwrite the v3-style config value.
+        set(result, parts, cssValue)
+      }
+
+      for (let part of parts) {
+        result = result?.[part]
+      }
+
+      return result
+    },
   }
 }
 
 export function registerPlugins(plugins: Plugin[], designSystem: DesignSystem, ast: AstNode[]) {
-  let pluginApi = buildPluginApi(designSystem, ast)
+  let pluginObjects = []
 
   for (let plugin of plugins) {
     if ('__isOptionsFunction' in plugin) {
       // Happens with `plugin.withOptions()` when no options were passed:
       // e.g. `require("my-plugin")` instead of `require("my-plugin")(options)`
-      plugin().handler(pluginApi)
+      pluginObjects.push(plugin())
     } else if ('handler' in plugin) {
       // Happens with `plugin(…)`:
       // e.g. `require("my-plugin")`
       //
       // or with `plugin.withOptions()` when the user passed options:
       // e.g. `require("my-plugin")(options)`
-      plugin.handler(pluginApi)
+      pluginObjects.push(plugin)
     } else {
       // Just a plain function without using the plugin(…) API
-      plugin(pluginApi)
+      pluginObjects.push({ handler: plugin, config: {} as UserConfig })
+    }
+  }
+
+  // Now merge all the configs and make all that crap work
+  let resolvedConfig = resolveConfig([
+    createCompatabilityConfig(designSystem.theme),
+    ...pluginObjects.map(({ config }) => config ?? {}),
+  ])
+
+  let pluginApi = buildPluginApi(designSystem, ast, resolvedConfig)
+
+  // Loop over the handlers and run them all with the resolved config + CSS theme probably somehow
+  for (let { handler } of pluginObjects) {
+    handler(pluginApi)
+  }
+}
+
+function readFromCss(theme: Theme, path: string) {
+  type ThemeValue =
+    // A normal string value
+    | string
+
+    // A nested tuple with additional data
+    | [main: string, extra: Record<string, string>]
+  // This only supports reading from the CSS Theme:…
+  let original = path
+    // Escape dots used inside square brackets
+    .replace(/\[(.*?)\]/g, (_, value) => `-${value.replace('.', '_')}`)
```

Comment:
While on the CSS `theme()` function I learned that this is valid syntax in V3: `theme(color[red].[500])`

Here's how I handled it, we might want to do something similar here I suppose: https://github.com/tailwindlabs/tailwindcss/pull/14177/files#diff-3e0c118787a4d539fc94ae61398cd2b7117ce62372be52991f873946336d3aa6R101-R126

Reply:
@philipp-spiess Did you find usage like this in the wild or was it just something that was working in v3 perhaps by coincidence and not by design?

Reply:
@adamwathan I found some unit tests that documented this behavior in V3: https://github.com/tailwindlabs/tailwindcss/blob/main/tests/evaluateTailwindFunctions.test.js#L68-L93

Let me see if I can find something in the wild.

Reply:
Didn't find anything in the open that was using more than one square bracket block in a `theme()` call but there are some that use brackets as separators e.g. `colors['blue-gray'].100`

Reply:
Ok that example isn't from a tailwind codebase lol

Reply:
> @adamwathan I found some unit tests that documented this behavior in V3: https://github.com/tailwindlabs/tailwindcss/blob/main/tests/evaluateTailwindFunctions.test.js#L68-L93
> 
> Let me see if I can find something in the wild.

Hmm that doesn't look the same to me, I don't see a dot between the square brackets there right?

```css
.banana {
  color: theme('colors[yellow][100]');
}
```

I dont see `].[` anywhere in that test file that I can find.